python/paddle/fluid/tests/unittests/test_matmul_op_output_fused.py

```diff
@@ -0,0 +1,145 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import print_function
+
+import unittest
+import numpy as np
+from paddle.fluid.tests.unittests.op_test import OpTest, skip_check_grad_ci
+from ir.inference.inference_pass_test import InferencePassTest
+
+import paddle.fluid as fluid
+
+
+@skip_check_grad_ci(reason="Tests inference only optimization.")
+class TestMatMulOpSpecial(OpTest):
+    def generate_data(self):
+        self.x = np.random.random([1, 128, 128]).astype("float32")
+        self.y = np.random.random([1, 128, 64]).astype("float32")
+        self.out = np.matmul(self.x, self.y)
+        self.shape_out = []
+        self.axis_out = []
+
+    def setUp(self):
+        self.op_type = "matmul"
+        self._cpu_only = True
+        self.use_mkldnn = True
+        self.generate_data()
+
+        self.inputs = {'X': self.x, 'Y': self.y}
+        self.attrs = {'use_mkldnn': self.use_mkldnn, }
+        if len(self.shape_out) > 0:
+            self.attrs['reshape_Out'] = self.shape_out
+            self.attrs['axis_Out'] = self.axis_out
+
+        self.outputs = {'Out': self.out}
+
+    def test_check_output(self):
+        self.check_output()
+
+
+@skip_check_grad_ci(reason="Tests inference only optimization.")
+class TestMatMulOpSpecialHackedXSimplest(TestMatMulOpSpecial):
+    def generate_data(self):
+        bs = 3
+        self.x = np.random.random([bs, 12, 128, 128]).astype("float32")
+        self.y = np.random.random([bs, 12, 128, 64]).astype("float32")
+        self.axis_out = [0, 2, 1, 3]
+        self.shape_out = [0, 0, 768]
+        self.out = np.matmul(self.x, self.y).transpose([0, 2, 1, 3]).reshape(
+            [bs, -1, 768])
+
+
+class TestMKLDNNMatmulFuseOp(InferencePassTest):
+    def setUp(self):
+        bs = 8
+        d_type = 'float32'
+        shape_x = [12, 128, 128]
+        shape_y = [12, 128, 64]
+        with fluid.program_guard(self.main_program, self.startup_program):
+            x = fluid.data(name='x', shape=[-1] + shape_x, dtype=d_type)
+            y = fluid.data(name='y', shape=[-1] + shape_y, dtype=d_type)
+            out = fluid.layers.matmul(x, y)
+            out = fluid.layers.transpose(out, perm=[0, 2, 1, 3])
+            out = fluid.layers.reshape(out, [0, 0, shape_y[0] * shape_y[2]])
+            out = fluid.layers.fc(out, size=1)
+
+        self.feeds = {
+            "x": np.random.random([bs] + shape_x).astype(d_type),
+            "y": np.random.random([bs] + shape_y).astype(d_type)
+        }
+        self.fetch_list = [out]
+        self.enable_mkldnn = True
+
+    def test_check_output(self):
+        use_gpu = False
+        self.check_output_with_option(use_gpu)
```

Comment:
I think the op with `reshape_out` and `transpose_out` , should not mix up with op without these two attributes set. But in this PR, I did not see key changes, so it means, when reuse, you may use wrong cached op, because you did not add attribute value to the key.

Could you also add a cache mechanism check for the fused op. Simply speaking, you just need to run 2 iterations. You can refer Michal's matmul PR, I think he did this work.

Reply:
Concerning mixing op without attributes.
The test you mentioned is doing two runs of the little topology from `setUp`. First is raw inference pass, the second is with MKLDNN on and than they are compared. There is no need to put the attributes `reshape_out` and `transpose_out` to MKLDNN MM here because they are modified implicitly during the fuse pass.
Concerning caching:
Thank you for the comment I'll check this

Reply:
@lidanqing-intel  OK I changed caching key and PADDLE_ENFORCE could you please see new version.

Reply:
Thanks for changing the key. LGTM.